### ISSUE TITLE
Techdocs  needs to send Identity information to the backend when requesting documentation and static files

### ISSUE THREAD

**Issue by GoWind** [backstage/backstage#4527]
Techdocs plugin currently does not provide authentication information to the backend or has an API to which this information
can be passed, when requesting Documentation and static files, such as images and CSS further requested by the Reader

## Feature Suggestion

1. Modify the TechDocs API  methods to optionally take in Authentication information
2. Techdocs fetches this info implicitly (from Identity or from `OAuth2` API) before making this request.
3. Reader rewrites links to static resources (CSS, images) to include authentication information.
## Possible Implementation

Modify `TechDocs` Implementation to request identity info during runtime and pass it along in all the methods in the  request
https://github.com/backstage/backstage/blob/master/plugins/techdocs/src/api.ts
or in 
https://github.com/backstage/backstage/blob/master/plugins/techdocs/src/reader/components/TechDocsPage.tsx
also in the docs links transformer possibly
https://github.com/backstage/backstage/blob/master/plugins/techdocs/src/reader/transformers/rewriteDocLinks.ts
## Context

Our Backstage application instance  is hosted on a publicly accessible endpoint and any request the app makes , to the backend instance, needs to be secure. When the frontend makes CORS requests to the backend, it is not clear who is request these routes/information thus causing a security vulnerability. 

In order to mitigate this, the TechDocs plugin should ideally provide the authentication/identity information so that the backend can authenticate the request and provide the requested  resource.

**Comment by Rugvip** [comment]
Tbh I think this extends beyond TechDocs for publicly hosted instances. You'd really want to protect the static content of the frontend as well. The only thing that wouldn't be protected would be the sign-in page, auth endpoints, and perhaps some other public content. Signing in would probably end up setting a session cookie for all endpoints, and if possible is also best handled by a piece of infrastructure that sits in front of the Backstage instance to harden things a bit more.

**Comment by GoWind** [comment]
Agreed. I assumed that as a part of this Big [PR](https://github.com/backstage/backstage/commit/cadceb2e1cd1f1cd59770a7488284128e1f0b4c1) , the Tech Docs would also be updated to  send `idToken` requests ,but I guess it wasn't covered. 

In any case, yes, we should find a way to secure static assets from the frontend as well. While they might not be valuable on their own, they can leak sensitive information about routes, components  via js module files and images, etc.

But this is a little tricky, as we need a way to signal to the frontend that the user has been authenticated , to set a session. This might involve no longer having a dumb webserver, but something that can accept a token, validate it with the generating authority (backstage-backend, possibly) and then set a session variable for the user. 

I would like to split the tasks, if possible as I am still pondering over what might be a good solution and how to solve this problem best. Atleast, initially, tech docs can be modified to send an `idToken` as with the other APIs, or some sort of a session. In the meantime , we can discuss ideas about how to secure the other static assets and possibly co-ordinate between the frontend and the backend to verify a token and generate a session.

**Comment by OrkoHunter** [comment]
In future, techdocs-backend would eventually want to know which user is sending the request and authenticate against a permissions service which knows if the requested user has access to the site. So +1 to this.

**Comment by GoWind** [comment]
Sounds good. Would like for me to introduce a PR for this (can assign the ticket to me) , along the links of https://github.com/backstage/backstage/commit/cadceb2e1cd1f1cd59770a7488284128e1f0b4c1 ? 
I can start thinking of adding identity information to `backend` requests first and then tackle the more general problem of securing the frontend's assets.

**Comment by OrkoHunter** [comment]
Sounds like a plan - awesome!

**Comment by OrkoHunter** [comment]
@GoWind Looks like @erikxiv just now created a PR https://github.com/backstage/backstage/pull/4542 which solves this!

**Comment by erikxiv** [comment]
FYI: As @GoWind pointed out in #4542, the PR fails to address static assets and thus won't work as-is :(

**Comment by erikxiv** [comment]
#4542 together with adding a cookie from adopter code as outlined in https://github.com/backstage/backstage/blob/master/contrib/docs/tutorials/authenticate-api-requests.md works. Not sure whether this satisfies this issue, or if the issue should be kept open awaiting a solution in backstage core to address iit.

**Comment by OrkoHunter** [comment]
Yeah, great work there @erikxiv. 👏 
@GoWind close it?

**Comment by OrkoHunter** [comment]
I believe this is okay to close.
